### PR TITLE
enable uds and fds for consul

### PR DESCRIPTION
**Pull request by ilackarms**
includes an e2e test against the petstore
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/905
resolves https://github.com/solo-io/gloo/issues/906
resolves https://github.com/solo-io/gloo/issues/913
resolves https://github.com/solo-io/gloo/issues/914

### CODE REVIEW

File: projects/gloo/pkg/plugins/consul/eds.go [hunk 1]
```diff
@@ -219,7 +221,9 @@ func buildEndpointName(service *consulapi.CatalogService) string {
 	if service.ServiceID != "" {
 		parts = append(parts, service.ServiceID)
 	}
-	return strings.Join(parts, "-")
+	unsanitizedName := strings.Join(parts, "-")
+	unsanitizedName = strings.ReplaceAll(unsanitizedName, "_", "")
```
Comment (kdorosh) [inline projects/gloo/pkg/plugins/consul/eds.go]: should this be a part of `kubeutils.SanitizeNameV2(unsanitizedName)`? or more likely, should is this a form of sanitization that we may want to move into this helper file?

File: projects/gloo/pkg/plugins/consul/plugin.go [hunk 2]
```diff
@@ -4,6 +4,9 @@ import (
 	"fmt"
 	"net/url"
 
+	"github.com/hashicorp/consul/api"
+	"github.com/pkg/errors"
```
Comment (EItanya) [inline projects/gloo/pkg/plugins/consul/plugin.go]: go-utils errors package

File: projects/gloo/pkg/plugins/consul/plugin.go [hunk 3]
```diff
@@ -26,10 +29,34 @@ func (p *plugin) Resolve(u *v1.Upstream) (*url.URL, error) {
...
+
+	// default to first datacenter
+	var dc string
+	if len(spec.DataCenters) > 0 {
+		dc = spec.DataCenters[0]
+	}
+
+	instances, _, err := p.client.Service(spec.ServiceName, "", &api.QueryOptions{Datacenter: dc, RequireConsistent: true})
+	if err != nil {
+		return nil, errors.Wrap(err, "getting service from catalog")
```
Comment (EItanya) [inline projects/gloo/pkg/plugins/consul/plugin.go]: The new error pardigm we've been trying to follow is defining errors as vars at the top of the file.
For example: https://github.com/solo-io/gloo/blob/master/projects/gloo/pkg/plugins/tcp/plugin.go

File: projects/gloo/pkg/plugins/consul/uds.go [hunk 4]
```diff
@@ -1,19 +1,83 @@
...
+}
+
+func UpdateUpstream(original, desired *v1.Upstream) (bool, error) {
+	originalSpec, ok := original.UpstreamSpec.UpstreamType.(*v1.UpstreamSpec_Consul)
+	if !ok {
+		return false, errors.Errorf("internal error: expected *v1.UpstreamSpec_Consul, got %v", reflect.TypeOf(original.UpstreamSpec.UpstreamType).Name())
+	}
+	desiredSpec, ok := desired.UpstreamSpec.UpstreamType.(*v1.UpstreamSpec_Consul)
+	if !ok {
+		return false, errors.Errorf("internal error: expected *v1.UpstreamSpec_Consul, got %v", reflect.TypeOf(original.UpstreamSpec.UpstreamType).Name())
```
Comment (EItanya) [inline projects/gloo/pkg/plugins/consul/uds.go]: Declare error above

File: projects/gloo/pkg/plugins/consul/eds.go (hunk 1 above)
Comment (ilackarms) [reply projects/gloo/pkg/plugins/consul/eds.go]: it's a bit of a special case here; if we replace `_` with `-`, the string will have multiple `-`s in a row (these are based on nomad task names which are prefixed with `_`)

not sure if we want to do a replace with `""` in the sanitize v2

**Pull request merged**